README.md
# Peatio - an open-source crypto currency exchange

### [peatio.tech](https://www.peatio.tech) fork

Peatio is a free and open-source crypto currency exchange implementation with the Rails framework and other cutting-edge technology.

## Mission

Our mission is to build the world best open-source crypto currency exchange with a high performance trading engine and safety which can be trusted and enjoyed by users. Additionally we want to move the crypto currency exchange technology forward by providing support and add new features. We are helping people to build easy their own exchange around the world.

Help is greatly appreciated, feel free to submit pull-requests or open issues.

## Things You Should Know

**RUNNING A EXCHANGE IS HARD.**

Peatio makes it easier, but running an exchange is still harder than a blog, which you can download the source code and following the guide or even a cool installer and boom!!! a fancy site is there to profit. We always prioritize security and speed higher than 1-click setup. We split Peatio to many components (processes) so it's flexible to deploy and scalable.

**SECURITY KNOWLEDGE IS A REQUIREMENT.**

Peatio cannot protect your customers when you leave your admin password 1234567, or open sensitive ports to public internet. No one can. Running an exchange is a very risky task because you're dealing with money directly. If you don't known how to make your exchange secure, hire an expert.

You must know what you're doing, there's no shortcut. Please get prepared before continue:

* Rails knowledge
* Security knowledge
* System administration

## Features

* Designed as high performance crypto currency exchange
* Built-in high performance matching-engine
* Built-in [Proof of Solvency](https://iwilcox.me.uk/2014/proving-bitcoin-reserves) Audit
* Built-in ticket system for customer support
* Usability and scalibility
* Websocket API and high frequency trading support
* Support multiple digital currencies (eg. Bitcoin, Litecoin, Dogecoin etc.)
* Easy customization of payment processing for both fiat and digital currencies
* SMS and Google Two-Factor authenticaton
* [KYC Verification](http://en.wikipedia.org/wiki/Know_your_customer)
* Powerful admin dashboard and management tools
* Highly configurable and extendable
* Industry standard security out of box
* Active community behind
* Free and open-source
* Maintained by [peatio.tech](https://www.peatio.tech)

## Known Exchanges using Peatio

* **[peatio.tech](https://www.peatio.tech)** - provides Peatio support (add custom features, deploy to the cloud, etc)
* [Yunbi Exchange](https://yunbi.com) - a crypto-currency exchange funded by BitFundPE
* [Bitspark](https://bitspark.io) - bitcoin exchange in Hong Kong
* [MarsX.io](https://acx.io) - australian cryptocurrency rxchange
* [One World Coin](https://oneworldcoin.com)

## Mobile Apps

* [Boilr](https://github.com/andrefbsantos/boilr) - cryptocurrency and bullion price alarms for Android

## Requirements

* Linux / Mac OSX
* Ruby 2.2.1
* Rails 4.0+
* Git 1.7.10+
* Redis 2.0+
* MySQL
* RabbitMQ

Find more details in the [docs directory](docs).

## Getting Started

Local development setup:

* [on Mac OS X](docs/dev/setup/osx.md)
* [on Ubuntu](docs/dev/setup/ubuntu.md)

Production setup:

* [Deploy production server on Ubuntu](docs/ops/deploy/ubuntu.md)
* [Using Peatio docker image](docs/ops/deploy/docker.md)

## API

You can interact with Peatio through API:

* [API v2 docs](https://demo.peatio.tech/documents/api_v2?lang=en)
* [Websocket API docs](https://demo.peatio.tech/documents/websocket_api)

Here're some API clients/wrappers:

* [peatio-client-ruby](https://github.com/peatio/peatio-client-ruby) is the official ruby client of both HTTP/Websocket API.
* [peatio-client-python by JohnnyZhao](https://github.com/JohnnyZhao/peatio-client-python) is a python client written by JohnnyZhao.
* [peatio-client-python by czheo](https://github.com/JohnnyZhao/peatio-client-python) is a python wrapper similar to peatio-client-ruby written by czheo.
* [peatioJavaClient](https://github.com/classic1999/peatioJavaClient.git) is a java client written by classic1999.
* [yunbi-client-php](https://github.com/panlilu/yunbi-client-php) is a php client written by panlilu.

## Custom Styles

Peatio front-end based Bootstrap 3.0 version and Sass, and you can custom exchange style for your mind.

* change bootstrap default variables in `vars/_bootstrap.css.scss`
* change peatio custom default variables in `vars/_basic.css.scss`
* add your custom variables in `vars/_custom.css.scss`
* add your custom css style in `layouts/_custom.css.scss`
* add or change features style in `features/_xyz.css.scss`

`vars/_custom.css.scss` can overwrite `vars/_basic.css.scss` defined variables
`layout/_custom.css.scss` can overwrite `layout/_basic.css.scss` and `layoputs/_header.css.scss` style

## Getting Involved

Want to report a bug, request a feature, contribute or translate Peatio?

* Browse our [issues](https://github.com/rubykube/peatio/issues),
  comment on proposals, report bugs.
* Clone the peatio repo, make some changes according to our development
  guidelines and issue a pull-request with your changes.
* If you need technical support or customization service,
  contact us: [hello@peatio.tech](mailto:hello@peatio.tech)

## Getting Support and Customization

If you need help with running/deploying/customizing Peatio,
you can contact us on [peatio.tech](https://www.peatio.tech).

## License

Peatio is released under the terms of the [MIT license](http://peatio.mit-license.org).

## What is Peatio?

[Peatio](http://en.wikipedia.org/wiki/Pixiu) (Chinese: 貔貅) is a Chinese mythical hybrid creature
considered to be a very powerful protector to practitioners of Feng Shui.

**[This illustration copyright for Peatio Team]**

![logo](public/peatio.png)


CHANGELOG.md
# Change Log

## [v0.2.5](https://github.com/rubykube/peatio/tree/v0.2.5) (2018-01-23)
[Full Changelog](https://github.com/rubykube/peatio/compare/0.2.4...v0.2.5)

**Fixed bugs:**

- Prevent ActiveRecord exceptions from being silenced  [\#295](https://github.com/rubykube/peatio/issues/295)
- Withdraws::Withdrawable\#create has became empty after removal of 2FA [\#290](https://github.com/rubykube/peatio/issues/290)
- Withdrawal of crypto \(both BTC & XRP\) not working [\#285](https://github.com/rubykube/peatio/issues/285)
- Issues with fees [\#280](https://github.com/rubykube/peatio/issues/280)
- Fiat deposits not working [\#276](https://github.com/rubykube/peatio/issues/276)
- Precompiled assets are broken at «Funds» page [\#253](https://github.com/rubykube/peatio/issues/253)
- Deletion of orders doesn't seem to work [\#248](https://github.com/rubykube/peatio/issues/248)
- Errors of data processing in loops may stop newer data for being processed \(stucks the whole application, daemons moustly\) [\#243](https://github.com/rubykube/peatio/issues/243)
- No job rescheduling in daemons when exception is raised [\#242](https://github.com/rubykube/peatio/issues/242)
- Generation of deposit address doesn't work stable: randomly isn't generated [\#241](https://github.com/rubykube/peatio/issues/241)
- Issue viewing documents in admin panel at «Verify Account» page [\#237](https://github.com/rubykube/peatio/issues/237)
- Missing translation at «Trade» [\#199](https://github.com/rubykube/peatio/issues/199)
- Button «Copy deposit address» at «Funds» is broken [\#197](https://github.com/rubykube/peatio/issues/197)
- Button «Reject» doesn't work while editing KYC document  [\#188](https://github.com/rubykube/peatio/issues/188)
- User should not be able to focus disabled fields when editing proof [\#186](https://github.com/rubykube/peatio/issues/186)
- Multiple admin menu rubrics receive "active" state as the same time [\#185](https://github.com/rubykube/peatio/issues/185)
- Missing translations for flash message in case member is disabled  [\#183](https://github.com/rubykube/peatio/issues/183)
- Error when you try to withdraw 0 from wallet with zero balance  [\#130](https://github.com/rubykube/peatio/issues/130)
- Spec is failing \(seed 13602\) [\#88](https://github.com/rubykube/peatio/issues/88)

**Closed issues:**

- CoinRPC\#sendtoaddress requires real number with maximum 8 decimal places but Peatio may send more \(withdraw\) [\#310](https://github.com/rubykube/peatio/issues/310)
- Write rake task which sends test email to specified address \(useful for testing mails\) [\#302](https://github.com/rubykube/peatio/issues/302)
- RabbitMQ doesn't requeue messages which consumer failed to process [\#298](https://github.com/rubykube/peatio/issues/298)
- Install exception reporting software which should be configurable [\#296](https://github.com/rubykube/peatio/issues/296)
- Remove sign in and sign up [\#291](https://github.com/rubykube/peatio/issues/291)
- Make SMTP username, password and authentication type optional  [\#271](https://github.com/rubykube/peatio/issues/271)
- Translations has been deleted but link still exists [\#269](https://github.com/rubykube/peatio/issues/269)
- Remove lib/tasks/migration.rake [\#268](https://github.com/rubykube/peatio/issues/268)
- Delete unneeded images & locales [\#267](https://github.com/rubykube/peatio/issues/267)
- Strip captcha feature [\#266](https://github.com/rubykube/peatio/issues/266)
- RabbitMQ server disconnects long-running clients [\#261](https://github.com/rubykube/peatio/issues/261)
- Fix markets sorting [\#260](https://github.com/rubykube/peatio/issues/260)
- Remove malformed currency symbol from title at page «Trade» [\#258](https://github.com/rubykube/peatio/issues/258)
- Delete locales leaving only English, French & Russian [\#256](https://github.com/rubykube/peatio/issues/256)
- Leftovers after acts-as-taggable-on removal [\#249](https://github.com/rubykube/peatio/issues/249)
- ChromeDriver constantly stucks [\#244](https://github.com/rubykube/peatio/issues/244)
- Leverage SMTP configuration using application.yml [\#239](https://github.com/rubykube/peatio/issues/239)
- Remove lib/tasks/deploy.rake [\#230](https://github.com/rubykube/peatio/issues/230)
- Remove lib/tasks/install.rake [\#226](https://github.com/rubykube/peatio/issues/226)
- Leftovers after removal of 2FA: member phone number & Phonelib [\#225](https://github.com/rubykube/peatio/issues/225)
- Remove running accounts [\#221](https://github.com/rubykube/peatio/issues/221)
- Remove signup history [\#220](https://github.com/rubykube/peatio/issues/220)
- Remove gem acts-as-taggable-on [\#219](https://github.com/rubykube/peatio/issues/219)
- Update OmniAuth & OmniAuth providers to the latest version  [\#216](https://github.com/rubykube/peatio/issues/216)
- Logs should go to stdout/stderr especially when running in Docker/Kubernetes [\#201](https://github.com/rubykube/peatio/issues/201)
- Leftovers after refactoring of CNY =\> CAD [\#200](https://github.com/rubykube/peatio/issues/200)
- Admin rubric "Documents" doesn't work at all [\#187](https://github.com/rubykube/peatio/issues/187)
- Add MailCatcher for testing mails in development environment [\#184](https://github.com/rubykube/peatio/issues/184)
- Remove leftovers after removal of 2fa auth [\#172](https://github.com/rubykube/peatio/issues/172)
- Update gem bunny [\#157](https://github.com/rubykube/peatio/issues/157)
- No way to customize RAILS\_ENV when building Docker image [\#126](https://github.com/rubykube/peatio/issues/126)
- Use $ as the default currency [\#67](https://github.com/rubykube/peatio/issues/67)
- Remove gem "unread" [\#40](https://github.com/rubykube/peatio/issues/40)
- Remove built-in ticketing system [\#22](https://github.com/rubykube/peatio/issues/22)
- Create a demo/test database with faker and factorybot [\#16](https://github.com/rubykube/peatio/issues/16)

**Merged pull requests:**


db/schema.rb
# encoding: UTF-8
# This file is auto-generated from the current state of the database. Instead
# of editing this file, please use the migrations feature of Active Record to
# incrementally modify your database, and then regenerate this schema definition.
#
# Note that this schema.rb definition is the authoritative source for your
# database schema. If you need to create the application database on another
# system, you should be using db:schema:load, not running all the migrations
# from scratch. The latter is a flawed and unsustainable approach (the more migrations
# you'll amass, the slower it'll run and the greater likelihood for issues).
#
# It's strongly recommended that you check this file into your version control system.

ActiveRecord::Schema.define(version: 20171130150354) do

  create_table "account_versions", force: true do |t|
    t.integer  "member_id"
    t.integer  "account_id"
    t.integer  "reason"
    t.decimal  "balance",         precision: 32, scale: 16
    t.decimal  "locked",          precision: 32, scale: 16
    t.decimal  "fee",             precision: 32, scale: 16
    t.decimal  "amount",          precision: 32, scale: 16
    t.integer  "modifiable_id"
    t.string   "modifiable_type"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "currency"
    t.integer  "fun"
  end

  add_index "account_versions", ["account_id", "reason"], name: "index_account_versions_on_account_id_and_reason", using: :btree
  add_index "account_versions", ["account_id"], name: "index_account_versions_on_account_id", using: :btree
  add_index "account_versions", ["member_id", "reason"], name: "index_account_versions_on_member_id_and_reason", using: :btree
  add_index "account_versions", ["modifiable_id", "modifiable_type"], name: "index_account_versions_on_modifiable_id_and_modifiable_type", using: :btree

  create_table "accounts", force: true do |t|
    t.integer  "member_id"
    t.integer  "currency"
    t.decimal  "balance",                         precision: 32, scale: 16
    t.decimal  "locked",                          precision: 32, scale: 16
    t.datetime "created_at"
    t.datetime "updated_at"
    t.decimal  "in",                              precision: 32, scale: 16
    t.decimal  "out",                             precision: 32, scale: 16
    t.integer  "default_withdraw_fund_source_id"
  end

  add_index "accounts", ["member_id", "currency"], name: "index_accounts_on_member_id_and_currency", using: :btree
  add_index "accounts", ["member_id"], name: "index_accounts_on_member_id", using: :btree

  create_table "api_tokens", force: true do |t|
    t.integer  "member_id",                        null: false
    t.string   "access_key",            limit: 50, null: false
    t.string   "secret_key",            limit: 50, null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "trusted_ip_list"
    t.string   "label"
    t.integer  "oauth_access_token_id"
    t.datetime "expires_at"
    t.string   "scopes"
    t.datetime "deleted_at"
  end

  add_index "api_tokens", ["access_key"], name: "index_api_tokens_on_access_key", unique: true, using: :btree
  add_index "api_tokens", ["secret_key"], name: "index_api_tokens_on_secret_key", unique: true, using: :btree

  create_table "assets", force: true do |t|
    t.string  "type"
    t.integer "attachable_id"
    t.string  "attachable_type"
    t.string  "file"
  end

  create_table "audit_logs", force: true do |t|
    t.string   "type"
    t.integer  "operator_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "auditable_id"
    t.string   "auditable_type"
    t.string   "source_state"
    t.string   "target_state"
  end

  add_index "audit_logs", ["auditable_id", "auditable_type"], name: "index_audit_logs_on_auditable_id_and_auditable_type", using: :btree
  add_index "audit_logs", ["operator_id"], name: "index_audit_logs_on_operator_id", using: :btree

  create_table "authentications", force: true do |t|
    t.string   "provider"
    t.string   "uid"
    t.string   "token"
    t.string   "secret"
    t.integer  "member_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "nickname"
  end

  add_index "authentications", ["member_id"], name: "index_authentications_on_member_id", using: :btree
  add_index "authentications", ["provider", "uid"], name: "index_authentications_on_provider_and_uid", using: :btree

  create_table "comments", force: true do |t|
    t.text     "content"
    t.integer  "author_id"
    t.integer  "ticket_id"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "deposits", force: true do |t|
    t.integer  "account_id"
    t.integer  "member_id"
    t.integer  "currency"
    t.decimal  "amount",                 precision: 32, scale: 16
    t.decimal  "fee",                    precision: 32, scale: 16
    t.string   "fund_uid"
    t.string   "fund_extra"
    t.string   "txid"
    t.integer  "state"
    t.string   "aasm_state"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.datetime "done_at"
    t.string   "confirmations"
    t.string   "type"
    t.integer  "payment_transaction_id"
    t.integer  "txout"
  end

  add_index "deposits", ["txid", "txout"], name: "index_deposits_on_txid_and_txout", using: :btree

  create_table "document_translations", force: true do |t|
    t.integer  "document_id", null: false
    t.string   "locale",      null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "title"
    t.text     "body"
    t.text     "desc"
    t.text     "keywords"
  end

  add_index "document_translations", ["document_id"], name: "index_document_translations_on_document_id", using: :btree
  add_index "document_translations", ["locale"], name: "index_document_translations_on_locale", using: :btree

  create_table "documents", force: true do |t|
    t.string   "key"
    t.string   "title"
    t.text     "body"
    t.boolean  "is_auth"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.text     "desc"
    t.text     "keywords"
  end

  create_table "fund_sources", force: true do |t|
    t.integer  "member_id"
    t.integer  "currency"
    t.string   "extra"
    t.string   "uid"
    t.boolean  "is_locked",  default: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.datetime "deleted_at"
  end

  create_table "id_documents", force: true do |t|
    t.integer  "id_document_type"
    t.string   "name"
    t.string   "id_document_number"
    t.integer  "member_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.date     "birth_date"
    t.text     "address"
    t.string   "city"
    t.string   "country"
    t.string   "zipcode"
    t.integer  "id_bill_type"
    t.string   "aasm_state"
  end

  create_table "identities", force: true do |t|
    t.string   "email"
    t.string   "password_digest"
    t.boolean  "is_active"
    t.integer  "retry_count"
    t.boolean  "is_locked"
    t.datetime "locked_at"
    t.datetime "last_verify_at"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "members", force: true do |t|
    t.string   "sn"
    t.string   "display_name"
    t.string   "email"
    t.integer  "identity_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "state"
    t.boolean  "activated"
    t.integer  "country_code"
    t.string   "phone_number"
    t.boolean  "disabled",     default: false
    t.boolean  "api_disabled", default: false
    t.string   "nickname"
  end

  create_table "oauth_access_grants", force: true do |t|
    t.integer  "resource_owner_id", null: false
    t.integer  "application_id",    null: false
    t.string   "token",             null: false
    t.integer  "expires_in",        null: false
    t.text     "redirect_uri",      null: false
    t.datetime "created_at",        null: false
    t.datetime "revoked_at"
    t.string   "scopes"
  end

  add_index "oauth_access_grants", ["token"], name: "index_oauth_access_grants_on_token", unique: true, using: :btree

  create_table "oauth_access_tokens", force: true do |t|
    t.integer  "resource_owner_id"
    t.integer  "application_id"
    t.string   "token",             null: false
    t.string   "refresh_token"
    t.integer  "expires_in"
    t.datetime "revoked_at"
    t.datetime "created_at",        null: false
    t.string   "scopes"
    t.datetime "deleted_at"
  end

  add_index "oauth_access_tokens", ["refresh_token"], name: "index_oauth_access_tokens_on_refresh_token", unique: true, using: :btree
  add_index "oauth_access_tokens", ["resource_owner_id"], name: "index_oauth_access_tokens_on_resource_owner_id", using: :btree
  add_index "oauth_access_tokens", ["token"], name: "index_oauth_access_tokens_on_token", unique: true, using: :btree

  create_table "oauth_applications", force: true do |t|
    t.string   "name",         null: false
    t.string   "uid",          null: false
    t.string   "secret",       null: false
    t.text     "redirect_uri", null: false
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "oauth_applications", ["uid"], name: "index_oauth_applications_on_uid", unique: true, using: :btree

  create_table "orders", force: true do |t|
    t.integer  "bid"
    t.integer  "ask"
    t.integer  "currency"
    t.decimal  "price",                     precision: 32, scale: 16
    t.decimal  "volume",                    precision: 32, scale: 16
    t.decimal  "origin_volume",             precision: 32, scale: 16
    t.integer  "state"
    t.datetime "done_at"
    t.string   "type",           limit: 8
    t.integer  "member_id"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "sn"
    t.string   "source",                                                            null: false
    t.string   "ord_type",       limit: 10
    t.decimal  "locked",                    precision: 32, scale: 16
    t.decimal  "origin_locked",             precision: 32, scale: 16
    t.decimal  "funds_received",            precision: 32, scale: 16, default: 0.0
    t.integer  "trades_count",                                        default: 0
  end

  add_index "orders", ["currency", "state"], name: "index_orders_on_currency_and_state", using: :btree
  add_index "orders", ["member_id", "state"], name: "index_orders_on_member_id_and_state", using: :btree
  add_index "orders", ["member_id"], name: "index_orders_on_member_id", using: :btree
  add_index "orders", ["state"], name: "index_orders_on_state", using: :btree

  create_table "partial_trees", force: true do |t|
    t.integer  "proof_id",   null: false
    t.integer  "account_id", null: false
    t.text     "json",       null: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "sum"
  end

  create_table "payment_addresses", force: true do |t|
    t.integer  "account_id"
    t.string   "address"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "currency"
  end

  create_table "payment_transactions", force: true do |t|
    t.string   "txid"
    t.decimal  "amount",                   precision: 32, scale: 16
    t.integer  "confirmations"
    t.string   "address"
    t.integer  "state"
    t.string   "aasm_state"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.datetime "receive_at"
    t.datetime "dont_at"
    t.integer  "currency"
    t.string   "type",          limit: 60
    t.integer  "txout"
  end

  add_index "payment_transactions", ["txid", "txout"], name: "index_payment_transactions_on_txid_and_txout", using: :btree
  add_index "payment_transactions", ["type"], name: "index_payment_transactions_on_type", using: :btree

  create_table "proofs", force: true do |t|
    t.string   "root"
    t.integer  "currency"
    t.boolean  "ready",                 default: false
    t.datetime "created_at"
    t.datetime "updated_at"
    t.string   "sum"
    t.text     "addresses"
    t.string   "balance",    limit: 30
  end

  create_table "read_marks", force: true do |t|
    t.integer  "readable_id"
    t.integer  "member_id",                null: false
    t.string   "readable_type", limit: 20, null: false
    t.datetime "timestamp"
  end

  add_index "read_marks", ["member_id"], name: "index_read_marks_on_member_id", using: :btree
  add_index "read_marks", ["readable_type", "readable_id"], name: "index_read_marks_on_readable_type_and_readable_id", using: :btree

  create_table "running_accounts", force: true do |t|
    t.integer  "category"
    t.decimal  "income",      precision: 32, scale: 16, default: 0.0, null: false
    t.decimal  "expenses",    precision: 32, scale: 16, default: 0.0, null: false
    t.integer  "currency"
    t.integer  "member_id"
    t.integer  "source_id"
    t.string   "source_type"
    t.string   "note"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "running_accounts", ["member_id"], name: "index_running_accounts_on_member_id", using: :btree
  add_index "running_accounts", ["source_id", "source_type"], name: "index_running_accounts_on_source_id_and_source_type", using: :btree

  create_table "signup_histories", force: true do |t|
    t.integer  "member_id"
    t.string   "ip"
    t.string   "accept_language"
    t.string   "ua"
    t.datetime "created_at"
  end

  add_index "signup_histories", ["member_id"], name: "index_signup_histories_on_member_id", using: :btree

  create_table "simple_captcha_data", force: true do |t|
    t.string   "key",        limit: 40
    t.string   "value",      limit: 6
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "simple_captcha_data", ["key"], name: "idx_key", using: :btree

  create_table "taggings", force: true do |t|
    t.integer  "tag_id"
    t.integer  "taggable_id"
    t.string   "taggable_type"
    t.integer  "tagger_id"
    t.string   "tagger_type"
    t.string   "context",       limit: 128
    t.datetime "created_at"
  end

  add_index "taggings", ["tag_id", "taggable_id", "taggable_type", "context", "tagger_id", "tagger_type"], name: "taggings_idx", unique: true, using: :btree

  create_table "tags", force: true do |t|
    t.string "name"
  end

  add_index "tags", ["name"], name: "index_tags_on_name", unique: true, using: :btree

  create_table "tickets", force: true do |t|
    t.string   "title"
    t.text     "content"
    t.string   "aasm_state"
    t.integer  "author_id"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  create_table "tokens", force: true do |t|
    t.string   "token"
    t.datetime "expires_at"
    t.integer  "member_id"
    t.boolean  "is_used",    default: false
    t.string   "type"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

  add_index "tokens", ["type", "token", "expires_at", "is_used"], name: "index_tokens_on_type_and_token_and_expires_at_and_is_used", using: :btree

  create_table "trades", force: true do |t|
    t.decimal  "price",         precision: 32, scale: 16
    t.decimal  "volume",        precision: 32, scale: 16
    t.integer  "ask_id"
    t.integer  "bid_id"
    t.integer  "trend"
    t.integer  "currency"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.integer  "ask_member_id"
    t.integer  "bid_member_id"
    t.decimal  "funds",         precision: 32, scale: 16
  end

  add_index "trades", ["ask_id"], name: "index_trades_on_ask_id", using: :btree
  add_index "trades", ["ask_member_id"], name: "index_trades_on_ask_member_id", using: :btree
  add_index "trades", ["bid_id"], name: "index_trades_on_bid_id", using: :btree
  add_index "trades", ["bid_member_id"], name: "index_trades_on_bid_member_id", using: :btree
  add_index "trades", ["created_at"], name: "index_trades_on_created_at", using: :btree
  add_index "trades", ["currency"], name: "index_trades_on_currency", using: :btree

  create_table "two_factors", force: true do |t|
    t.integer  "member_id"
    t.string   "otp_secret"
    t.datetime "last_verify_at"
    t.boolean  "activated"
    t.string   "type"
    t.datetime "refreshed_at"
  end

  create_table "versions", force: true do |t|
    t.string   "item_type",  null: false
    t.integer  "item_id",    null: false
    t.string   "event",      null: false
    t.string   "whodunnit"
    t.text     "object"
    t.datetime "created_at"
  end

  add_index "versions", ["item_type", "item_id"], name: "index_versions_on_item_type_and_item_id", using: :btree

  create_table "withdraws", force: true do |t|
    t.string   "sn"
    t.integer  "account_id"
    t.integer  "member_id"
    t.integer  "currency"
    t.decimal  "amount",     precision: 32, scale: 16
    t.decimal  "fee",        precision: 32, scale: 16
    t.string   "fund_uid"
    t.string   "fund_extra"
    t.datetime "created_at"
    t.datetime "updated_at"
    t.datetime "done_at"
    t.string   "txid"
    t.string   "aasm_state"
    t.decimal  "sum",        precision: 32, scale: 16, default: 0.0, null: false
    t.string   "type"
  end

end
